### PR TITLE
Impro 1669 part2

### PR DESCRIPTION
**Pull request by Kat-90**
Addresses #GitHubissuenum

Description:
Updated unit tests under the circular kernel folder. Subtask related to 1669.

Testing:
 - [X] Ran tests and they passed OK
 - [ ] Added new tests for the new feature(s)

CLA
 - [ ] If a new developer, signed up to CLA

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  31
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py [hunk 1]
```diff
@@ -44,6 +45,10 @@
 from improver.nbhood.circular_kernel import (
     GeneratePercentilesFromACircularNeighbourhood,
 )
+from improver.synthetic_data.set_up_test_cubes import (
+    add_coordinate,
+    set_up_variable_cube,
+)
 
 from ..nbhood.test_BaseNeighbourhoodProcessing import set_up_cube, set_up_cube_lat_long
```
Comment (MoseleyS) [inline improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:53]: Unused imports

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  31
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Geneal comment, as discussed, it would be better to change the cube definition strategy so that setUp defines a cube of ones and each test sets the zero points relevant to that test. It'll be less confusing. I think you inherited a rather confused strategy in these tests!

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 2]
```diff
@@ -38,6 +38,7 @@
 from iris.tests import IrisTest
 
 from improver.nbhood.circular_kernel import CircularNeighbourhood
+from improver.synthetic_data.set_up_test_cubes import set_up_variable_cube
 
 from ..nbhood.test_BaseNeighbourhoodProcessing import (
```
[the remark this answers is not in the document]
Comment (Kat-90) [reply improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:43]: Done.

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 3]
```diff
@@ -77,11 +78,20 @@ class Test_apply_circular_kernel(IrisTest):
...
+
     def test_basic(self):
         """Test that the plugin returns an iris.cube.Cube."""
-        cube = set_up_cube(
-            zero_point_indices=((0, 0, 2, 2),), num_time_points=1, num_grid_points=5
-        )
+
+        data = np.ones((5, 5), dtype=np.float32)
+        data[2, 2] = 0
+        cube = set_up_variable_cube(data, spatial_grid="equalarea",)
```
[the remark this answers is not in the document]
Comment (Kat-90) [reply improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:93]: Done.

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py (hunk 1 above)
Comment (Kat-90) [reply improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:53]: Done.

**Comment by Kat-90** [comment]
I'm inclined to keep the set_up_cube zero points in the test_make_percentile_cube class, as these are never changed in the individual tests, but all others have been set in their tests.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  26
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  26
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

File: improver_tests/acceptance/SHA256SUMS [hunk 4]
```diff
@@ -139,7 +139,6 @@ a0f4aece2816a7e4c51e85da223c18feac4c3ec583ba59e8dd7893ec4f7efcae  ./generate-met
 801bbcb93345faab177b94546f2b5fc5f4585e59b10e6e9a680fe1f600b0b906  ./generate-metadata-cube/kgo_realization.nc
 9223d25967889c999583533e4994fba1c2b1dfc8030e91b506dc9b72778e5567  ./generate-metadata-cube/kgo_single_height_level.nc
 ca37652252852f04bc3016f9c0172cafdebbacabfd8d1a906780b7c4126a8f62  ./generate-metadata-cube/kgo_variable_cube_json_inputs.nc
-6e74662c6236549d17d58bb709970a383f7fae05fadd9cda7a79ab97012e8339  ./generate-metadata-cube/leading_dimensions.json
```
Comment (MoseleyS) [inline improver_tests/acceptance/SHA256SUMS:142]: This shouldn't happen. Try "git merge upstream master"

File: improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py [hunk 5]
```diff
@@ -77,11 +77,16 @@ class Test_apply_circular_kernel(IrisTest):
 
     """Test neighbourhood circular probabilities plugin."""
 
+    def setUp(self):
+        """Set up a cube."""
+
+        data = np.ones((1, 16, 16), dtype=np.float32)
+        self.cube = set_up_variable_cube(data, spatial_grid="equalarea",)
```
Comment (MoseleyS) [inline improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py:84]: It now occurs to me that we have a scalar dimension on these cubes (realization). This shouldn't occur in IMPROVER, so I thinkn we should change all these to remove scalar dimensions. It'll make all the tests slightly simpler too.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
This looks better. Thanks Kat. A couple more thoughts for you.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py  3
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  33
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py  3
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  33
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

**Review by MoseleyS — DISMISSED** [review]

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py [hunk 6]
```diff
@@ -146,10 +156,12 @@ def test_2d_slice(self):
...
                     [1.0, 1.0, 1.0, 1.0, 1.0],
                 ],
-            ]
+            ],
         )
         kernel = np.array([[0.0, 1.0, 0.0], [1.0, 1.0, 1.0], [0.0, 1.0, 0.0]])
-        cube = self.cube[0, 0, :, :]
+        cube = self.cube
+        cube.data[2, 2] = 0
+        cube = cube[:, :]
```
Comment (cgsandford) [inline improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:164]: Not sure what these three statements are doing.  What would happen if you just had `self.cube.data[2, 2] = 0`, and then called the plugin on `self.cube`?

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py [hunk 7]
```diff
@@ -181,10 +193,12 @@ def test_irregular_kernel(self):
...
                     [1.0, 1.0, 1.0, 1.0, 1.0],
                 ],
-            ]
+            ],
         )
         kernel = np.array([[0.0, 1.0, 0.0], [1.0, 0.0, 1.0], [0.0, 0.0, 1.0]])
-        cube = self.cube[0, 0, :, :]
+        cube = self.cube
+        cube.data[2, 2] = 0
+        cube = cube[:, :]
```
Comment (cgsandford) [inline improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:201]: As above

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py [hunk 8]
```diff
@@ -193,10 +207,14 @@ def test_irregular_kernel(self):
...
-        )  # Just within range of the edge.
-        slice_2d = cube[0, 0, :, :]
+
+        data = np.ones((3, 3), dtype=np.float32)
+        data[1, 1] = 0
+        cube = set_up_variable_cube(data, spatial_grid="equalarea",)
+
+        # Just within range of the edge.
+
+        slice_2d = cube[:, :]
```
Comment (cgsandford) [inline improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:217]: The cube is already 2D.  What effect does this statement have?

**Review by cgsandford — CHANGES_REQUESTED** [review]
Two minor points that come up repeatedly, so rather than go through every line I'm just going to state them here:

- There are a lot of `cube = self.cube` statements in these tests.  You can get rid of these completely, and then manipulate and operate on `self.cube` directly.  The reason for this is that `setUp` is automatically called to reinitialise `self.cube` for each test case - so changing the data in `self.cube` in one test doesn't affect any subsequent tests.  (Actually you are doing that anyway, because python doesn't copy unless explicitly told to do so - but that's an aside we can discuss offline if you're interested!)
- There are a number of `cube = cube[:, :]` statements that I'm not sure are needed any more.  Could you have a look at what happens without these, please (see comment below for exactly what I'm looking for)?

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py (hunk 6 above)
Comment (Kat-90) [reply improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:164]: Done. Yes, for some reason I didn't think self.cube.data was liked so I did it in stages but it is completely fine so I've changed it all now.

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py (hunk 8 above)
Comment (Kat-90) [reply improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:217]: No effect now, apologies for the oversight, it used to be a 3D cube with a realization of 1 which was useless so I removed it.

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py (hunk 7 above)
Comment (Kat-90) [reply improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:201]: Done.

File: improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py [hunk 9]
```diff
@@ -67,49 +71,48 @@ class Test_make_percentile_cube(IrisTest):
...
+
         result = GeneratePercentilesFromACircularNeighbourhood().make_percentile_cube(
-            cube
+            self.cube
         )
         self.assertIsInstance(result.coord("percentile"), iris.coords.Coord)
         self.assertArrayEqual(result.coord("percentile").points, DEFAULT_PERCENTILES)
-        self.assertArrayEqual(result[0].data, cube.data)
-        self.assertDictEqual(cube.metadata._asdict(), result.metadata._asdict())
+        self.assertArrayEqual(result[0].data, self.cube.data)
```
Comment (cgsandford) [inline improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py:97]: Can I be pedantic here and ask that you create an expected data array before running the processing.  That would look like:
```
expected_data = self.cube.data.copy()
...
self.assertArrayEqual(result[0].data, expected_data)
```
That catches a case where the input cube is inadvertently modified.

**Review by cgsandford — CHANGES_REQUESTED** [review]
As discussed, could you make sure you've made the changes in the `test_CircularNeighbourhood` file too please, and regenerate the correct checksums.  Just one other comment below.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py  3
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  33
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- improver_tests/nbhood/circular_kernel/test_CircularNeighbourhood.py  3
- improver_tests/nbhood/circular_kernel/test_GeneratePercentilesFromACircularNeighbourhood.py  33
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6181810&bid=20176502)

**Review by cgsandford — APPROVED** [review]

**Pull request merged**